src/ExtMsg.actor.cpp

```diff
@@ -626,41 +626,36 @@ ACTOR Future<Reference<IReadWriteContext>> insertDocument(Reference<CollectionCo
                                                           bson::BSONObj d,
                                                           Optional<IdInfo> encodedIds) {
 	state Standalone<StringRef> encodedId;
-	state Standalone<StringRef> valueEncodedId;
 	state Optional<bson::BSONObj> idObj;
 	state Reference<QueryContext> dcx;
 
 	if (!encodedIds.present()) {
-		encodedId = DataValue(bson::OID::gen()).encode_key_part();
-		valueEncodedId = encodedId;
-		idObj = Optional<bson::BSONObj>();
+		// No _id field, generate a new one.
+		state bson::OID generatedId = bson::OID::gen();
+		state bson::BSONObj docWithGeneratedId =
+		    bson::BSONObjBuilder().appendElements(d).append("_id", generatedId).obj();
+		encodedId = DataValue(generatedId).encode_key_part();
 		dcx = cx->cx->getSubContext(encodedId);
+		// It's unlikely that the auto generated ID collides with an existing one, but just to be safe.
+		Optional<DataValue> existing = wait(dcx->get(DataValue("_id", DVTypeCode::STRING).encode_key_part()));
+		if (existing.present()) {
+			TraceEvent(SevError, "insertDocument: Auto generated object id collides with an existing ID")
+			    .detail("Auto generated ID", generatedId.toString());
+			throw duplicated_key_field();
+		}
+		// Insert doc first
+		insertElementRecursive("", docWithGeneratedId, dcx);
 	} else {
 		encodedId = encodedIds.get().keyEncoded;
-		valueEncodedId = encodedIds.get().valueEncoded;
-		idObj = encodedIds.get().objValue;
 		dcx = cx->cx->getSubContext(encodedId);
+
 		Optional<DataValue> existing = wait(dcx->get(DataValue("_id", DVTypeCode::STRING).encode_key_part()));
```

Comment:
This is a nice cleanup. There is still more duplicate code between `if` and `else` blocks. Can you remove all the duplication entirely? We could just have `if` and `else` blocks just to set variables.